narwhal/executor/src/subscriber.rs

```diff
@@ -138,11 +137,9 @@ impl<Network: SubscriberNetwork> Subscriber<Network> {
                 },
 
                 // Receive here consensus messages for which we have downloaded all transactions data.
-                (message, permit) = join(waiting.next(), self.tx_notifier.reserve()), if !waiting.is_empty() => {
-                    if let Ok(permit) = permit {
-                        permit.send(message.expect("We don't poll empty queue"));
-                    } else {
-                        error!("tx_notifier closed");
+                message = waiting.next(), if !waiting.is_empty() => {
```

Comment:
Do we need now the check `if !waiting.is_empty()` ? I believe we can remove it

Reply:
We do want it, otherwise `waiting.next` resolves right away and we get busy loop

Reply:
What do you mean busy loop? It will resolve to an `Option` right? Having something like:
```
Some(message) = waiting.next()
```

wouldn't be pretty much similar to
```
message = waiting.next(), if !waiting.is_empty()
```
?

Reply:
Ok, got it, I think you are right: https://github.com/MystenLabs/sui/pull/5139